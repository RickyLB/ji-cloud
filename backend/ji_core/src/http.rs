use std::net::{SocketAddr, TcpListener};

/// Creates a [`TcpListener`] via `listenfd` if possible.
///
/// [`TcpListener`]: https://doc.rust-lang.org/stable/std/net/struct.TcpListener.html
pub fn get_tcp_fd() -> Option<TcpListener> {
    #[cfg(feature = "listenfd")]
    let fd = listenfd::ListenFd::from_env().take_tcp_listener(0).unwrap();

    #[cfg(not(feature = "listenfd"))]
    let fd = None;

    fd
}

/// Get the port to run the server on.
pub fn get_addr(default: Option<u16>) -> SocketAddr {
    // 0 is a special port that means "assign me whatever port you want"
    let port = std::env::var("PORT")
        .ok()
        .and_then(|it| it.parse().ok())
        .or(default)
        .unwrap_or(0);

    ([0, 0, 0, 0], port).into()
}
