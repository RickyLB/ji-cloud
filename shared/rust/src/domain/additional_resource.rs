//! Types for additional resources for JIG or Courses.

use crate::domain::{
    asset::AssetId, audio::AudioId, image::ImageId, meta::ResourceTypeId, pdf::PdfId,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Wrapper type around [`Uuid`](Uuid), represents the ID of an additional resource.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
pub struct AdditionalResourceId(pub Uuid);

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
/// Over-the-wire representation of a JIG or Course additional resource.
pub struct AdditionalResource {
    /// The additional resources's ID.
    pub id: AdditionalResourceId,

    /// Name for additional resource
    pub display_name: String,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Content of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

/// Request to create a new `AdditionalResource`.
///
/// [`additional_resource::Create`](crate::api::endpoints::additional_resource::Create)
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceCreateRequest {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    pub asset_id: AssetId,

    /// Display name for additional resource
    pub display_name: String,

    /// Type of additional resource
    pub resource_type_id: ResourceTypeId,

    /// Value of additional resource
    #[serde(flatten)]
    pub resource_content: ResourceContent,
}

/// Request to update an `AdditionalResource`.
///
/// [`additional_resource::Update`](crate::api::endpoints::additional_resource::Update)
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalResourceUpdateRequest {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset_id: Option<AssetId>,

    /// resource display name
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// Type of additional  resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub resource_type_id: Option<ResourceTypeId>,

    /// Kind of additional resource
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(flatten)]
    pub resource_content: Option<ResourceContent>,
}

/// Delete an `AdditionalResource` by Asset Id.
///
/// [`additional_resource::GetLive`](crate::api::endpoints::additional_resource::GetLive)
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct AssetIdResource {
    /// Asset Id (JIG or Course) for additional resource
    #[serde(flatten)]
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub asset_id: Option<AssetId>,
}

/// Value of additional resource
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ResourceContent {
    /// Additional resource kind: image
    ImageId(ImageId),
    /// Additional resource kind: audioFile
    AudioId(AudioId),
    /// Additional resource kind: link
    Link(url::Url),
    /// Additional resource kind: pdf
    PdfId(PdfId),
}

into_uuid![AdditionalResourceId];
