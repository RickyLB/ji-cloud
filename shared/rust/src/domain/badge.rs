//! Types for Badges.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::asset::UserOrMe;

/// Wrapper type around [`Uuid`], represents the ID of a Badge.
#[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
#[cfg_attr(feature = "backend", derive(sqlx::Type))]
#[cfg_attr(feature = "backend", sqlx(transparent))]
pub struct BadgeId(pub Uuid);

/// The response returned when a request for `GET`ing a Badge is successful.
#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Badge {
    /// The ID of the Badge.
    pub id: BadgeId,

    /// The Badge's name.
    pub display_name: String,

    /// Creator of Badge
    pub created_by: Uuid,

    /// Description of the Badge. Defaults to empty string.
    pub description: String,

    /// Number of members on Badge
    pub member_count: u32,

    /// Thumbnail URL
    pub thumbnail: url::Url,

    /// When Badge was created
    pub created_at: DateTime<Utc>,

    /// When Badge was last edited
    pub last_edited: Option<DateTime<Utc>>,
}

/// Request to create a new Badge.
///
/// This creates the draft and live [Badge Data](Badge Data) copies with the requested info.
#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BadgeCreateRequest {
    /// The Badge's name.
    pub display_name: String,

    /// Description of the Badge. Defaults to empty string.
    pub description: String,

    /// Thumbnail of the Badge
    pub thumbnail: url::Url,
}

/// Request for updating a Badge's draft data.
#[derive(Serialize, Deserialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct BadgeUpdateRequest {
    /// The Badge's name to be updated.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub display_name: Option<String>,

    /// The current author to be updated
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub creator_id: Option<Uuid>,

    /// Description of the Badge to be updated.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub description: Option<String>,

    /// Thumbnail of the Badge to be updated.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub thumbnail: Option<url::Url>,
}

/// Query for [`Browse`](crate::api::endpoints::badge::Browse).
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct BadgeBrowseQuery {
    /// creator of badges
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator_id: Option<UserOrMe>,

    /// The page number to get.
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page: Option<u32>,

    /// The hits per page to be returned
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_limit: Option<u32>,
}

/// Response for [`Browse`](crate::api::endpoints::badge::Browse).
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BadgeBrowseResponse {
    /// the Badges returned.
    pub badges: Vec<Badge>,

    /// The number of pages found.
    pub pages: u32,

    /// The total number of Badges found
    pub total_badge_count: u64,
}

/// Search for Badges via the given query string.
#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct BadgeSearchQuery {
    /// The query string.
    #[serde(default)]
    #[serde(skip_serializing_if = "String::is_empty")]
    pub q: String,

    /// The page number of the Badges to get.
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page: Option<u32>,

    /// Optionally filter by author's id
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator_id: Option<UserOrMe>,

    /// Optionally filter by the author's name
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator_name: Option<String>,

    /// The hits per page to be returned
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_limit: Option<u32>,
}

/// Response for [`Search`](crate::api::endpoints::badge::Search).
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct BadgeSearchResponse {
    /// the Badges returned.
    pub badges: Vec<Badge>,

    /// The number of pages found.
    pub pages: u32,

    /// The total number of Badges found
    pub total_badge_count: u64,
}

/// Members associated with Badge
#[derive(Serialize, Deserialize)]
pub struct BrowseMembersResponse {
    /// user id of member
    pub members: Vec<Uuid>,

    /// user id of member
    pub count: u32,
}

into_uuid![BadgeId];
