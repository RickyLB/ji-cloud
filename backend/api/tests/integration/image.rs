mod recent;
mod tag;
mod user;

use http::StatusCode;
use macros::test_service;
use serde_json::json;
use shared::domain::{image::ImageId, CreateResponse};
use sqlx::postgres::{PgConnectOptions, PgPoolOptions};
use uuid::Uuid;

use crate::{
    fixture::Fixture,
    helpers::{setup_service, LoginExt},
    service::Service,
};

async fn create(
    styles: &[Uuid],
    age_ranges: &[Uuid],
    affiliations: &[Uuid],
    categories: &[Uuid],
    tags: &[i16],
    port: u16,
) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": (),
            "styles": styles,
            "age_ranges": age_ranges,
            "affiliations": affiliations,
            "categories": categories,
            "tags": tags,
            "size": "Canvas",
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: CreateResponse<ImageId> = resp.json().await?;

    insta::assert_json_snapshot!(body, {".id" => "[id]"});

    Ok(())
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::Image", "Fixture::MetaKinds")
)]
async fn create_no_meta(port: u16) -> anyhow::Result<()> {
    create(&[], &[], &[], &[], &[], port).await
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::Image", "Fixture::MetaKinds")
)]
async fn create_with_styles(port: u16) -> anyhow::Result<()> {
    create(
        &[
            "6389eaa0-de76-11ea-b7ab-0399bcf84df2".parse()?,
            "6389ff7c-de76-11ea-b7ab-9b5661dd4f70".parse()?,
        ],
        &[],
        &[],
        &[],
        &[],
        port,
    )
    .await
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::Image", "Fixture::MetaKinds")
)]
async fn create_with_meta(port: u16) -> anyhow::Result<()> {
    create(
        &["6389eaa0-de76-11ea-b7ab-0399bcf84df2".parse()?],
        &["f3722790-de76-11ea-b7ab-77b45e9af3ef".parse()?],
        &["c0cd4446-de76-11ea-b7ab-93987e8aa112".parse()?],
        &[],
        &[1],
        port,
    )
    .await
}

async fn create_error(kind: &str, id: &str, port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": null,
            "styles": [],
            "age_ranges": [],
            "affiliations": [],
            "categories": [],
            "tags": [],
            "size": "Canvas",
            kind: [id],
        }))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::UNPROCESSABLE_ENTITY);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}

async fn create_error_tag(kind: &str, id: &i16, port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .login()
        .json(&json!({
            "name": "test",
            "description": "testest",
            "is_premium": false,
            "publish_at": null,
            "styles": [],
            "age_ranges": [],
            "affiliations": [],
            "categories": [],
            "tags": [],
            "size": "Canvas",
            kind: [id],
        }))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::UNPROCESSABLE_ENTITY);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[test_service(setup = "setup_service", fixtures("Fixture::User"))]
async fn create_with_style_error(port: u16) -> anyhow::Result<()> {
    create_error("styles", "6389eaa0-de76-11ea-b7ab-0399bcf84df2", port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User"))]
async fn create_with_affiliation_error(port: u16) -> anyhow::Result<()> {
    create_error("affiliations", "6389eaa0-de76-11ea-b7ab-0399bcf84df2", port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User"))]
async fn create_with_age_range_error(port: u16) -> anyhow::Result<()> {
    create_error("age_ranges", "6389eaa0-de76-11ea-b7ab-0399bcf84df2", port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User"))]
async fn create_with_category_error(port: u16) -> anyhow::Result<()> {
    create_error("categories", "6389eaa0-de76-11ea-b7ab-0399bcf84df2", port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User"))]
async fn create_with_tags_error(port: u16) -> anyhow::Result<()> {
    create_error_tag("tags", &22, port).await
}

#[test_service(
    setup = "setup_service",
    fixtures(
        "Fixture::User",
        "Fixture::Image",
        "Fixture::MetaKinds",
        "Fixture::MetaImage"
    )
)]
async fn get_metadata(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

// todo: search; missing algolia, s3
// todo: upload; missing algolia, s3
// todo: delete; missing algolia, s3
// todo: delete: edge case (never uploaded, should work even without s3), missing algolia

async fn update(req: &serde_json::Value, port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .json(req)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/image/3095d05e-f2c7-11ea-89c3-3b621dd74a1f",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::MetaKinds", "Fixture::Image")
)]
async fn update_empty(port: u16) -> anyhow::Result<()> {
    update(&json!({}), port).await
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::MetaKinds", "Fixture::Image")
)]
async fn update_is_premium(port: u16) -> anyhow::Result<()> {
    update(&json!({"is_premium": true}), port).await
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::MetaKinds", "Fixture::Image")
)]
async fn update_styles(port: u16) -> anyhow::Result<()> {
    update(&json!({"styles": ["6389eaa0-de76-11ea-b7ab-0399bcf84df2", "6389ff7c-de76-11ea-b7ab-9b5661dd4f70"]}), port
    )
    .await
}

#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::MetaKinds", "Fixture::Image")
)]
async fn update_tags(port: u16) -> anyhow::Result<()> {
    update(&json!({"tags": [0, 2]}), port).await
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Image"))]
async fn browse(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    // create a new image resource
    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/image/browse", port))
        .query(&[("page", "0"), ("size", "Canvas")])
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".metadata.updated_at" => "[timestamp]"});

    Ok(())
}

// https://cloud.google.com/storage/docs/performing-resumable-uploads#single-chunk-upload
#[ignore]
#[test_service(
    setup = "setup_service",
    fixtures("Fixture::User", "Fixture::Image"),
    services("Service::GoogleCloudStorage")
)]
async fn upload_with_url(port: u16) -> anyhow::Result<()> {
    let file: Vec<u8> = include_bytes!("../../fixtures/images/ji-logo.png").to_vec();

    let client = reqwest::Client::new();

    // get an upload URL for the new media
    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/image/8cca720a-c4bb-11eb-8edf-63da1d86939c/raw",
            port,
        ))
        .json(&json!({ "file_size": &file.len() }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let resp: shared::domain::image::ImageUploadResponse = resp.json().await?;

    let resp = client
        .put(&resp.session_uri)
        .header(reqwest::header::CONTENT_LENGTH, &file.len().to_string())
        .body(file)
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::OK);

    Ok(())
}

#[ignore]
#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Image"))]
async fn create_media_and_upload_with_url(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    // create a new image resource
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/image", port))
        .json(&json!({
            "name": "test_img",
            "description": "test_descip",
            "is_premium": false,
            "publish_at": (),
            "styles": [],
            "age_ranges": [],
            "affiliations": [],
            "tags": [],
            "categories": [],
            "size": "Canvas",
        }))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(StatusCode::CREATED, resp.status());

    let resp: shared::domain::image::CreateResponse = resp.json().await?;

    let id = resp.id.0;

    // get an upload URL for the new media
    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/image/{}/raw",
            port,
            id.to_string()
        ))
        .json(&json!({"file_size": 5}))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let resp: shared::domain::image::ImageUploadResponse = resp.json().await?;

    let _url = resp.session_uri;

    // perform upload in single chunk
    let _resp = client;

    Ok(())
}
