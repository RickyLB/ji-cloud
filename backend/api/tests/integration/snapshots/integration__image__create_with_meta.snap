---
source: tests/integration/image.rs
expression: body
---
{
  "id": "[id]"
}
